types/mapcoins.go

```diff
@@ -0,0 +1,42 @@
+package types
+
+// map coins is a map representation of sdk.Coins
+// intended solely for use in bulk additions.
+// All serialization and iteration should be done after conversion to sdk.Coins.
+type MapCoins map[string]Int
+
+func NewMapCoins(coins Coins) MapCoins {
+	m := make(MapCoins, len(coins))
+	m.Add(coins...)
+	return m
+}
+
+func (m MapCoins) Add(coins ...Coin) {
+	for _, coin := range coins {
+		existAmt, exists := m[coin.Denom]
```

Comment:
Since `ToCoins` skips zero-amount coins, wouldn't it be more efficient to skip adding coins that are zero-amount in the first place? Since negative amounts aren't allowed, a non-zero amount can't drop back down to zero.